firedrake/mesh.py

```diff
@@ -1036,6 +1041,10 @@ def __init__(self, coordinates):
 
         self._coordinates = coordinates
 
+        # Cached coordinate functions in different function spaces
+        # dict: FunctionSpace -> (coordinate Function, interpolator)
+        self.aux_coordinate_functions = {}
```

Comment:
Can this map from ufl Element to values, rather than functionspace? Also, can it be a `weakref.WeakValueDictionary` to the `interpolator`. There must be (or should be) a way to get the target function out of an interpolator object. This way we don't introduce ref-cycles.

Reply:
Oh, I see that the user makes the weakrefs.

Reply:
So presumably it should be the "parent" ufl Element, i.e. `fs.ufl_element().sub_elements()[0]` in case `fs` is a vector function space?

Reply:
I've updated this now, storing the interpolator only